hubgrep/frontend_blueprint/forms/hosting_service/hosting_services.py

```diff
@@ -0,0 +1,142 @@
+import json
+import logging
+
+import markdown
+
+from flask_wtf import FlaskForm
+from wtforms import StringField, SelectField, TextAreaField
+from wtforms.widgets.html5 import URLInput
+from wtforms.validators import ValidationError, URL
+from wtforms.validators import InputRequired
+
+from hubgrep.lib.hosting_service_interfaces import hosting_service_interface_mapping
+from hubgrep.models import HostingService
+
+from hubgrep.frontend_blueprint.forms.hosting_service.validators import (
+    validate_custom_config,
+    validate_url,
+    validate_api_url,
+)
+
+from hubgrep.frontend_blueprint.forms.hosting_service.hosting_service_first_step import HostingServiceFormFirstStep
+
+logger = logging.getLogger(__name__)
+
+
+class NoHostingServiceFormException(Exception):
+    pass
+
+class HostingServiceForm(HostingServiceFormFirstStep):
+    """ Final step when adding a new hosting-service, doubling as the form used for editing hosting-services. """
+
+    # validate_url sanitizes url before validate_api_url makes a call to it
+    api_url = StringField(
+        "Api Url",
+        [InputRequired(), URL(), validate_url, validate_api_url],
+        widget=URLInput(),
+        description="should be the same as the landing page in most cases"
+    )
+    custom_config = TextAreaField(
+        "Custom config",
+        [validate_custom_config],
+        description="Custom config json. If you dont know what to put in, just leave it empty.",
+    )
+    important_notes_md = ""
+
+    def populate_api_url(self):
+        self.api_url.data = self.landingpage_url.data
+
+    def to_hosting_service(self):
+        """
+        create hosting service (the db model, not the interface) from form data
+        """
+        h = HostingService()
+
+        # set current_user later - this should be request-agnostic :)
+        h.user_id = None
+
+        h.api_url = self.api_url.data
+        h.landingpage_url = self.landingpage_url.data
+        h.type = self.type.data
+        if hasattr(self, "api_key"):
+            h.api_key = self.api_key.data
+        else:
+            h.api_key = None
+        h.custom_config = self.custom_config.data
```

Comment:
If you want to delete the custom_config, that seems OK with - it just seems like something very likely to be useful to have set up already (hubgrep_search -> frontend form -> hubgrep_indexer) - its a bit of wiring to get it back later, plus database migrations.